aten/src/ATen/core/thread_pool.h

```diff
@@ -11,17 +15,81 @@ namespace ivalue {
 struct Future;
 } // namespace ivalue
 
-struct C10_API ThreadPool final {
-  ThreadPool();
+class C10_API TaskThreadPoolBase {
+ public:
+  virtual void run(const std::function<void()>& func) = 0;
 
-  ~ThreadPool();
+  virtual size_t size() const = 0;
 
-  void schedule(std::function<void(void)> task);
+  /**
+   * The number of available (i.e. idle) threads in this thread pool.
+   */
+  virtual size_t numAvailable() const = 0;
 
-  void workOnTasksUntilCompleted(c10::intrusive_ptr<ivalue::Future> future);
+  virtual ~TaskThreadPoolBase();
+};
+
+class C10_API ThreadPool : public c10::TaskThreadPoolBase {
+ protected:
+  struct task_element_t {
+    bool run_with_id;
+    const std::function<void()> no_id;
+    const std::function<void(std::size_t)> with_id;
+
+    explicit task_element_t(const std::function<void()>& f)
+        : run_with_id(false), no_id(f), with_id(nullptr) {}
+    explicit task_element_t(const std::function<void(std::size_t)>& f)
+        : run_with_id(true), no_id(nullptr), with_id(f) {}
+  };
+
+  std::queue<task_element_t> tasks_;
+  std::vector<std::thread> threads_;
+  std::mutex mutex_;
+  std::condition_variable condition_;
+  std::condition_variable completed_;
+  bool running_;
+  bool complete_;
+  std::size_t available_;
+  std::size_t total_;
+  int numa_node_id_;
+
+ public:
+  ThreadPool() = delete;
+
+  explicit ThreadPool(std::size_t pool_size, int numa_node_id = -1);
+
+  ~ThreadPool() override;
+
+  size_t size() const override;
+
+  size_t numAvailable() const override;
+
+  /// @brief Add task to the thread pool if a thread is currently available.
+  template <typename Task>
```

Comment:
What is this templated? It immediately casts the task to `std::function<void()>`? @ilia-cher do you remember?

Reply:
Seems safe to remove. I will leave it for @ilia-cher to justify.

Reply:
check the usages but yes we don't seem to be using it